### PR TITLE
Ring structure on tuples

### PR DESCRIPTION
**Pull request by pleroy**
Paddock calls!  (Macbeth, I, 1)

### CODE REVIEW

File: geometry/cartesian_product.hpp [hunk 1]
```diff
@@ -37,11 +37,23 @@ struct CartesianProductAdditiveGroup;
...
+// This class exports one function:
+//
+//   static constexpr ... Multiply(LTuple const& left, RTuple const& right);
+//
+// where the return types are appropriate for the operation.  The element types
+// of the tuples must have a structure of multiplicative ring.
+template<typename LTuple, typename RTuple,
+         int lsize_ = std::tuple_size_v<LTuple>,
+         int rsize_ = std::tuple_size_v<RTuple>>
+class CartesianProductRing;
```
Comment (eggrobin) [inline geometry/cartesian_product.hpp]: CartesianProductAdditiveGroup and CartesianProductVectorSpace make sense as names (they are the direct product of groups and direct product of vector spaces respectively), but this has to be called PolynomialRing; it is emphatically *not* the product ring.

File: geometry/cartesian_product.hpp [hunk 2]
```diff
@@ -37,11 +37,23 @@ struct CartesianProductAdditiveGroup;
...
          typename = std::make_integer_sequence<int, std::tuple_size_v<Tuple>>>
-struct CartesianProductVectorSpace;
+class CartesianProductVectorSpace;
+
+// This class exports one function:
+//
+//   static constexpr ... Multiply(LTuple const& left, RTuple const& right);
+//
+// where the return types are appropriate for the operation.  The element types
+// of the tuples must have a structure of multiplicative ring.
```
Comment (eggrobin) [inline geometry/cartesian_product.hpp]: "The element types of the tuples must have a structure of multiplicative ring" is not a good description of the requirements (it sounds like the requirement for the product ring, which this is not).

File: geometry/cartesian_product.hpp (hunk 1 above)
Comment (pleroy) [reply geometry/cartesian_product.hpp]: Done.

File: geometry/cartesian_product.hpp (hunk 2 above)
Comment (pleroy) [reply geometry/cartesian_product.hpp]: Done.

File: geometry/cartesian_product.hpp [hunk 3]
```diff
@@ -21,19 +21,30 @@ template<typename LTuple, typename RTuple>
...
 constexpr auto operator/(Tuple const& left, Scalar const& right);
 
 }  // namespace cartesian_product
+
+namespace polynomial_ring {
+
+template<typename LTuple, typename RTuple,
+         typename = std::enable_if_t<quantities::is_tuple_v<LTuple>>,
+         typename = std::enable_if_t<quantities::is_tuple_v<RTuple>>>
+constexpr auto operator*(LTuple const& left, RTuple const& right);
```
Comment (eggrobin) [inline geometry/cartesian_product.hpp]: Should there be a comment that the cartesian product is interpreted as the coordinates in the monomial basis (or should it say monomial basis in the namespace name)?

Comment (pleroy) [reply geometry/cartesian_product.hpp]: Added a comment.

**Pull request merged**